packages/@uppy/aws-s3/src/index.ts

```diff
@@ -385,10 +386,59 @@ export default class AwsS3Multipart<
     return this.#client
   }
 
+  #setClient(opts?: Partial<AwsS3MultipartOptions<M, B>>) {
+    if (
+      opts == null ||
+      !(
+        'endpoint' in opts ||
+        'companionUrl' in opts ||
+        'headers' in opts ||
+        'companionHeaders' in opts ||
+        'cookiesRule' in opts ||
+        'companionCookiesRule' in opts
+      )
+    )
+      return
+    if ('companionUrl' in opts && !('endpoint' in opts)) {
+      this.uppy.log(
+        '`companionUrl` option has been removed in @uppy/aws-s3, use `endpoint` instead.',
+        'warning',
+      )
+    }
+    if ('companionHeaders' in opts && !('headers' in opts)) {
+      this.uppy.log(
+        '`companionHeaders` option has been removed in @uppy/aws-s3, use `headers` instead.',
+        'warning',
+      )
+    }
+    if ('companionCookiesRule' in opts && !('cookiesRule' in opts)) {
+      this.uppy.log(
+        '`companionCookiesRule` option has been removed in @uppy/aws-s3, use `cookiesRule` instead.',
+        'warning',
+      )
+    }
+    if ('endpoint' in opts) {
+      this.#client = new RequestClient(this.uppy, {
+        pluginId: this.id,
+        provider: 'AWS',
+        companionUrl: this.opts.endpoint!,
+        companionHeaders: this.opts.headers,
+        companionCookiesRule: this.opts.cookiesRule,
+      })
+    } else {
+      if ('headers' in opts) {
+        this.#setCompanionHeaders()
+      }
+      if ('cookiesRule' in opts) {
+        this.#client.opts.companionCookiesRule = opts.cookiesRule
+      }
+    }
+  }
+
   setOptions(newOptions: Partial<AwsS3MultipartOptions<M, B>>): void {
     this.#companionCommunicationQueue.setOptions(newOptions)
-    super.setOptions(newOptions)
-    this.#setCompanionHeaders()
+    super.setOptions(newOptions as any)
```

Comment:
something wrong here with the types?
```suggestion
    super.setOptions(newOptions as any)
```

Reply:
```
Argument of type 'Partial<AwsS3MultipartOptions<M, B>>' is not assignable to parameter of type 'Partial<((DefinePluginOpts<AwsS3MultipartOptions<M, B>, "allowedMetaFields" | "limit" | "retryDelays" | "getTemporarySecurityCredentials" | "shouldUseMultipart"> & Pick<...> & Required<...>) & Partial<...>) & AWSS3MultipartWithoutCompanionMandatorySignPart<...> & AWSS3NonMultipartWithoutCompanionMandatory<...>>'.
  Type 'Partial<_AwsS3MultipartOptions & AWSS3WithoutCompanion & AWSS3NonMultipartWithoutCompanionMandatory<M, B> & { ...; }>' is not assignable to type 'Partial<((DefinePluginOpts<AwsS3MultipartOptions<M, B>, "allowedMetaFields" | "limit" | "retryDelays" | "getTemporarySecurityCredentials" | "shouldUseMultipart"> & Pick<...> & Required<...>) & Partial<...>) & AWSS3MultipartWithoutCompanionMandatorySignPart<...> & AWSS3NonMultipartWithoutCompanionMandatory<...>>'.
```

Reply:
yikes! i don't know how to fix that

Reply:
We can always improve this in a follow-up PR, I don't think it's blocking

Reply:
agree